receiver/apachedruidreceiver/metrics.go

```diff
@@ -0,0 +1,308 @@
+	mReceiver.logger.Debug("starting receiver HTTP server")
+	// We use l.server.Serve* over l.server.ListenAndServe*
+	// So that we can catch and return errors relating to binding to network interface on start.
+	var listenConfig net.ListenConfig
+
+	listener, err := listenConfig.Listen(ctx, "tcp", mReceiver.config.Endpoint)
+	if err != nil {
+		return err
+	}
+
+	mReceiver.wg.Add(1)
+	if mReceiver.config.TLS != nil {
+		go func() {
+			defer mReceiver.wg.Done()
+
+			mReceiver.logger.Debug("Starting ServeTLS",
+				zap.String("address", mReceiver.config.Endpoint),
+				zap.String("certfile", mReceiver.config.TLS.CertFile),
+				zap.String("keyfile", mReceiver.config.TLS.KeyFile))
+
+			err := mReceiver.server.ServeTLS(listener, mReceiver.config.TLS.CertFile, mReceiver.config.TLS.KeyFile)
+
+			mReceiver.logger.Debug("Serve TLS done")
+
+			if err != http.ErrServerClosed {
+				mReceiver.logger.Error("ServeTLS failed", zap.Error(err))
+				host.ReportFatalError(err)
+			}
+		}()
+	} else {
+		go func() {
+			defer mReceiver.wg.Done()
+
+			mReceiver.logger.Debug("Starting Serve",
+				zap.String("address", mReceiver.config.Endpoint))
+
+			err = mReceiver.server.Serve(listener)
+
+			mReceiver.logger.Debug("Serve done")
+
+			if err != http.ErrServerClosed {
+				mReceiver.logger.Error("Serve failed", zap.Error(err))
+				host.ReportFatalError(err)
+			}
+		}()
+	}
+
+	return nil
+}
+
+func (mReceiver *metricsReceiver) handleRequest(rw http.ResponseWriter, request *http.Request) {
+	if mReceiver.config.BasicAuth != nil {
+		auth := request.Header.Get("Authorization")
+		credentials := fmt.Sprintf("%s:%s", mReceiver.config.BasicAuth.Username, mReceiver.config.BasicAuth.Password)
+		configAuth := fmt.Sprintf("Basic %s", base64.StdEncoding.EncodeToString([]byte(credentials)))
+		if auth == "" {
+			rw.WriteHeader(http.StatusUnauthorized)
+			mReceiver.logger.Debug("Got request with no basic auth credentials when they were specified in config, dropping...")
+			return
+		} else if auth != configAuth {
+			rw.WriteHeader(http.StatusUnauthorized)
+			mReceiver.logger.Debug("Got request with incorrect basic auth credentials when they were specified in config, dropping...")
+			return
+		}
+	}
+
+	if request.Method != "POST" {
+		rw.WriteHeader(http.StatusBadRequest)
+		mReceiver.logger.Debug("Receiver server only accepts POST requests", zap.String("remote", request.RemoteAddr))
+		return
+	}
+
+	if request.Header.Get("Content-Type") != "application/json" {
+		rw.WriteHeader(http.StatusBadRequest)
+		mReceiver.logger.Debug("Content type must be JSON", zap.String("remote", request.RemoteAddr))
+		return
+	}
+
+	payload, err := io.ReadAll(request.Body)
+	if err != nil {
+		rw.WriteHeader(http.StatusBadRequest)
+		mReceiver.logger.Debug("Failed to read metrics payload", zap.Error(err), zap.String("remote", request.RemoteAddr))
+		return
+	}
+
+	var metrics []interface{}
```

Comment:
Is there a reason why we decode to an interface if we probably know what the data is supposed to look like? It'd be good to use composition here so we don't have so many interfaces throughout the code as they are unwieldy and 

I think we should Unmarshall into an expected struct that we expect the data to look like 

```go
type Datapoint struct {
  Name string `json:"metric"`
  Value float64 `json:"value"`
}

```

And then unmarshalling becomes a tad more simple:

```go
var metrics []Datapoint{}
	if err = json.Unmarshal(payload, &metrics); err != nil {
                ...
		return
	}
```